### ISSUE TITLE
Sequential vs Concurrent execution

### ISSUE THREAD

**Issue by erichulburd** [sharkdp/hyperfine#58]
I looked over the documentation, but, unless I'm missing something, I don't see an option for running commands sequentially versus concurrently (ie for load testing). If there is no option for specifying this, I'm assuming that commands run sequentially, is this correct?

Thanks.

**Comment by sharkdp** [comment]
That is correct. All benchmark runs are performed sequentially.

**Comment by drewwells** [comment]
Can we reopen this to add support for concurrent executions?

**Comment by sharkdp** [comment]
Yes, I can reopen this. Could you please describe your use case in more detail?

**Comment by davinkevin** [comment]
I'm really interested about this too. 

To describe my need, I'm using hyperfine for performance testing on some endpoint of application, during development. My underlying stack is fully reactive, but we have to take care of blocking thread. So if I run only 1 call at a time, I won't reach the limit of thread allowed (even if I do bad thing in my code). The result in hyperfine will be great and I will continu my development like if everything is ok.

But, in reactive development, the performance drops when we reach a number of concurrent/parallels request at the same time. Because if we don't take care of our thread limitation (or call things which block threads) we will exhaust the thread pool … and the performance will drastically reduce. 

For this use case, I think it could be good to provide a way to make concurrent calls from hyperfine.

**Comment by muxator** [comment]
My use case is simpler, but I think allowing parallel execution could be beneficial.

I am comparing the performance of various versions of the same command line application (there are 4 "switches" that can be independently toggled on/off, thus leading to 16 different cases).

The application is single threaded and performs no network and disk access. I am running each test a number of times (let's say 5) to check for consistency. This results in a total of 80 runs, each several minutes long.

My machine is "beefy" (this statement makes no sense in absolute terms, but is reasonable in this specific case): it has 32 cores and 64 GB of RAM. If I ran with 32 simultaneous processes, my results would not be reliable, because I'd probably end up being thermally limited. However, I am confident that running 2-4 tests in parallel would not hamper the quality of my results too much, while giving a good boost in terms of wall clock time.

An attention point is that only advanced users would know how to safely enable concurrency without making their measurements unreliable: this is clearly a double edged sword.

I do not know if the development effort for implementing this is too much compared to the expected benefit, but I wanted to share a possible use case anyway.

Thanks for this wonderful piece of software!

**Comment by divinity76** [comment]
i too want to benchmark parallel runs; 

as a shitty workaround, it's possible to write
```
hyperfine 'yes "" | head -n5 | xargs -P5 -L1 ./prog'
```
which should run ./prog 5 times in parallel, and hyperfine measures how long it took to run all 5 invocations in parallel... i think
(there's probably a better way to write that too, it was just the first method that came to mind; most likely GNU Parallel should be used, but i don't know how to use GNU Parallel)

**Comment by divinity76** [comment]
i too want to benchmark parallel runs;

as a shitty workaround, it's possible to write

hyperfine 'yes "" | head -n5 | parallel ./prog'

which should run ./prog 5 times in parallel, and hyperfine measures how long it took to run all 5 invocations in parallel...

**Comment by Maple38** [comment]
Why was this removed from the Hyperfine 2.0 milestone? Would be massively useful.

**Comment by hochej** [comment]
First of all, thank you for hyperfine. I would like to add a use case that has not been mentioned yet.

I use hyperfine on a high-performance computing (HPC) cluster. As background information, all benchmarked tasks run on compute nodes rather than on the login node where I launch hyperfine. I use hyperfine to verify that performance is consistent across different nodes in our cluster, and also to conduct scaling studies to understand how applications perform with varying numbers of CPU cores or GPUs.

For this use case, concurrent execution would be extremely valuable. Since each benchmark runs on an independent compute node, there's no resource contention between them. For example, when testing how an application scales from 1 to 32 cores, I currently have to wait for each configuration to complete before moving to the next, even though I have dozens of idle nodes available.

In my workflow, we use [Slurm](https://slurm.schedmd.com/overview.html) to allocate resources on specific nodes as part of the `--prepare` step, then release them with `--cleanup`. The SSH overhead to launch jobs is negligible compared to the actual computation time. Having concurrent execution would significantly reduce the wall-clock time for comprehensive performance tests, especially when each individual benchmark might take several minutes and I need to test many different configurations.

This feature would make hyperfine really interesting for HPC/supercomputing environments where parallel resource availability is the common.

**Comment by rbeneyton** [comment]
First of all, thank you for hyperfine.

As I really needed this concurrent execution feature, and interlaced processes, I coded a small hyperfine-like tool for my needs: [megafine](https://crates.io/crates/megafine) (yes the name means nothing but crates.io namespace is almost full, and it's a reference to this one).
I initially attempted to implement that feature inside hyperfine but this is really intrusive so I understand @sharkdp postponement.

If that can unblock some of you (@erichulburd , @drewwells , maybe @davinkevin , @muxator , @divinity76 and @hochej ), don't hesitate to test if that fits your requirements too.
